xdsl/dialects/linalg.py

```diff
@@ -490,6 +496,112 @@ def __init__(
+    result = var_result_def(AnyTensorType)
+
+    permutation = attr_def(DenseArrayBase)
+
+    def __init__(
+        self,
+        input: SSAValue,
+        init: SSAValue,
+        permutation: Attribute,
+        result: Attribute | None = None,
+    ):
+        super().__init__(
+            attributes={
+                "permutation": permutation,
+            },
+            operands=(input, init),
+            result_types=(result,),
+        )
+
+    def verify_(self) -> None:
+
+        assert isinstance(input_type := self.input.type, TensorType | MemRefType)
+        assert isinstance(init_type := self.init.type, TensorType | MemRefType)
+
+        input_shape = input_type.get_shape()
+        init_shape = init_type.get_shape()
+
+        if (input_rank := len(input_shape)) != (init_rank := len(init_shape)):
+            raise VerifyException(
+                f"Input rank ({input_rank}) does not match init rank ({init_rank})"
+            )
+        if (input_rank := len(input_shape)) != (
+            permutation_size := len(self.permutation.data)
+        ):
+            raise VerifyException(
+                f"Input rank ({input_rank}) does not match size of permutation ({permutation_size})"
+            )
+
+        permutation_shape = cast(list[int], self.permutation.as_tuple())
+
+        for i in range(len(input_shape)):
+            input_dimension = input_shape[permutation_shape[i]]
+            init_dimension = init_shape[i]
+
+            if input_dimension != init_dimension:
+                raise VerifyException(
+                    f"dim(result, {i}) = {init_dimension} "
+                    f"doesn't match dim(input, permutation[{i}]) = {input_dimension}"
+                )
+
+    def print(self, printer: Printer):
+        printer.print_string(" ins(")
+        printer.print(self.input)
+        printer.print_string(":")
+        printer.print(self.input.type)
+        printer.print_string(")")
+        printer.print_string(" outs(")
+        printer.print(self.init)
+        printer.print_string(":")
+        printer.print(self.init.type)
+        printer.print_string(") ")
+        printer.print_string("permutation")
+        printer.print_string(" = ")
+        printer.print(list(self.permutation.as_tuple()))
+
+    @classmethod
+    def parse(cls, parser: Parser) -> Self:
+        parser.parse_characters("ins")
+        parser.parse_punctuation("(")
+        input = parser.parse_operand()
+        parser.parse_punctuation(":")
+        parser.parse_type()
+        parser.parse_punctuation(")")
+        parser.parse_characters("outs")
+        parser.parse_punctuation("(")
+        init = parser.parse_operand()
+        parser.parse_punctuation(":")
+        # result type the same type as init
+        result = parser.parse_type()
+        parser.parse_punctuation(")")
+        parser.parse_keyword("permutation")
+        parser.parse_punctuation("=")
+        permutation = parser.parse_comma_separated_list(
```

Comment:
out of curiosity, what is the parsed type when this line is just `permutation = parser.parse_attribute()`? I'd expect it to parse an array of i64

Reply:
` Operation does not verify: [1 : i64, 0 : i64] should be of base attribute array`
it parses an array yes, but it needs to be a dense array

Reply:
kk, thank you